8328894: HttpResponse.body() returns null with https target and failed proxy authentication

ehs208 opened the pull request:
### Summary
  This patch addresses the issue where `HttpResponse.body()` returns `null` for 407 responses when using HTTPS through a proxy, while HTTP requests correctly return the response body.

  ### Problem
  When an HTTPS request receives a 407 Proxy Authentication Required response, the response body is discarded during CONNECT tunnel establishment. This is inconsistent with HTTP behavior where the body is properly returned.

  **Root cause:**
  - HTTPS uses `MultiExchange<Void>` for CONNECT requests
  - The body is explicitly ignored via `ignoreBody()` on 407 responses
  - No mechanism exists to preserve the body for later retrieval

  ### Proposed Solution
  I propose the following changes to preserve and return the 407 response body:

  1. **PlainTunnelingConnection.java**: Change `MultiExchange<Void>` to `MultiExchange<byte[]>` and read the body on 407 responses instead of ignoring it

  2. **ProxyAuthenticationRequired.java**: Add `proxyResponseBody` field to carry the body bytes through the exception

  3. **Exchange.java**: Cache both the proxy response and body, then return them when the application calls `body()`

  ### Testing
  Added comprehensive test (`ProxyAuthHttpTest.java`) covering:
  - Basic HTTP and HTTPS 407 responses
  - Multiple `BodyHandler` types: `ofString()`, `ofByteArray()`, `ofInputStream()`, `ofLines()`
  - Response headers validation

  **Test results**: 38/38 passed

  ### Notes
  - This change only affects 407 responses; all other flows remain unchanged
  - The cached body is cleared after first use to prevent reuse
  - No changes to public APIs; internal implementation only

  I'd appreciate any feedback on this approach. If there's a better way to handle this, I'm happy to revise.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- \[ \] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- \[x\] Change must not contain extraneous whitespace
- \[x\] Commit message must refer to an issue

### Error
&nbsp;⚠️ Pull request body is missing required line: `- [x] I confirm that I make this contribution in accordance with the [OpenJDK Interim AI Policy](https://openjdk.org/legal/ai).`

### Issue
 * [JDK-8328894](https://bugs.openjdk.org/browse/JDK-8328894): HttpResponse.body() returns null with https target and failed proxy authentication (**Bug** - P3)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/28232/head:pull/28232` \
`$ git checkout pull/28232`

Update a local copy of the PR: \
`$ git checkout pull/28232` \
`$ git pull https://git.openjdk.org/jdk.git pull/28232/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 28232`

View PR using the GUI difftool: \
`$ git pr show -t 28232`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/28232.diff">https://git.openjdk.org/jdk/pull/28232.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/28232#issuecomment-3516465257)
</details>

ehs208 committed:
8328894: Read 407 response body in HTTPS CONNECT tunneling

ehs208 committed:
8328894: Add proxyResponseBody field to preserve 407 body

ehs208 committed:
8328894: Cache and return 407 response body to application

ehs208 committed:
8328894: Add test for 407 response body with various handlers

ehs208 committed:
Merge branch 'master' into JDK-8328894

ehs208 committed:
fix whitespace issue

ehs208 commented:
/touch

ehs208 commented:
/touch

vy commented on src/java.net.http/share/classes/jdk/internal/net/http/Exchange.java:
```diff
@@ -226,6 +230,30 @@ public CompletableFuture<T> readBodyAsync(HttpResponse.BodyHandler<T> handler) {
+                    }
+                    public void cancel() {}
+                });
+                return subscriber.getBody().toCompletableFuture();
```
Doesn't this need to be `MinimalFuture.of(subscriber.getBody())`?

vy replied:
Superseded by [this comment](https://github.com/openjdk/jdk/pull/28232#discussion_r3719399060).

ehs208 commented:
/touch

vy commented on src/java.net.http/share/classes/jdk/internal/net/http/Exchange.java:
```diff
@@ -27,6 +27,8 @@
 
 import java.io.IOException;
 import java.net.ProtocolException;
+import java.net.http.HttpClient.Version;
+import java.net.http.HttpHeaders;
```
Unused imports.

I suggest reverting all cosmetic/styling/redundant changes, in particular, the ones in `PlainTunnelingConnection`. They create noise in the VCS history and make reviews difficult.

vy replied:
Superseded by [this comment](https://github.com/openjdk/jdk/pull/28232#discussion_r3719451162).

vy commented on src/java.net.http/share/classes/jdk/internal/net/http/ProxyAuthenticationRequired.java:
```diff
@@ -34,16 +34,16 @@
+        this(proxyResponse, null);
+    }
+
+    public ProxyAuthenticationRequired(Response proxyResponse, byte[] proxyResponseBody) {
```
This is an internal class, we don't need to preserve backward compatibility, we can just remove the old `ProxyAuthenticationRequired#new(Response)` ctor.

ehs208 commented:
/touch

ehs208 commented:
/touch

vy commented:
@ehs208, I see you're keeping the issue alive, though I see no further commits. I've dropped some remarks, but got no response so far. Are you intending to work on this in the future, or are you waiting for some other review feedback?

ehs208 commented:
Hi @vy, I do intend to continue working on this. The RDP1 fork was completed some time ago, and I had previously asked whether I should proceed with an updated patch or keep this on hold.

I can't find the remarks you mentioned. Could you point me to them?

vy commented on src/java.net.http/share/classes/jdk/internal/net/http/Exchange.java:
```diff
@@ -453,14 +481,20 @@ public CompletableFuture<Response> responseAsync() {
+            cachedProxyBody = par.proxyResponseBody;
+            // Don't set bodyIgnored if we have a cached body
+            if (cachedProxyBody == null) {
+                bodyIgnored = MinimalFuture.completedFuture(null);
```
`CompletableFuture<Void> bodyIgnored` is essentially never used, except being treated as a boolean flag. It can very well be just a `boolean`. Plus, we now have 3 `volatile` state variables:

1. `bodyIgnored` (which is effectively equivalent to `cachedProxyBody == null`)
2. `cachedProxyBody`
3. `cachedProxyResponse`

I share the following proposals for your consideration:

- [Collapse 3 states into 1](https://github.com/user-attachments/files/30744775/01-collapse-states.patch)
- [Fix the custom `Flow.Subscription`](https://github.com/user-attachments/files/30744118/02-improve-subscription.patch)

vy commented on src/java.net.http/share/classes/jdk/internal/net/http/PlainTunnelingConnection.java:
This change set is super excessive. It is impossible to see you actually changed 3-4 LoC. Consider [this patch reducing your changes to bare minimum](https://github.com/user-attachments/files/30737808/03-simplify-PTC.patch).

vy commented on test/jdk/java/net/httpclient/ProxyAuthHttpTest.java:
IMHO, this test can benefit a rewrite. See [this proposal](https://github.com/user-attachments/files/30742742/04-simplify-test.patch).

vy commented:
@ehs208, how would you like to proceed with this issue? If it is okay with you, I can also pick it up, and, of course, have you added as a contributor.

ehs208 commented:
> @ehs208, how would you like to proceed with this issue? If it is okay with you, I can also pick it up, and, of course, have you added as a contributor.

Hi @vy, I'd like to continue with this myself. I'm working on an updated commit now that addresses your feedback, plus a couple of additional issues I found while reworking it. Will push shortly.
